lib/internal/modules/esm/loader.js

```diff
@@ -802,36 +828,42 @@ class ESMLoader {
       hookErrIdentifier: '',
       hookIndex: chain.length - 1,
       hookName: 'resolve',
+      isChainAsync: false,
       shortCircuited: false,
     };
-
     const context = {
       conditions: DEFAULT_CONDITIONS,
       importAssertions,
       parentURL,
     };
-    const validate = (hookErrIdentifier, { 0: suppliedSpecifier, 1: ctx }) => {
 
+    const validateArgs = (hookErrIdentifier, { 0: suppliedSpecifier, 1: ctx }) => {
       validateString(
         suppliedSpecifier,
         `${hookErrIdentifier} specifier`,
       ); // non-strings can be coerced to a url string
 
       validateObject(ctx, `${hookErrIdentifier} context`);
     };
+    const validateOutput = (hookErrIdentifier, output) => {
+      if (
+        typeof output !== 'object' || // [2]
+        typeof output.then === 'function'
```

Comment:
Do we have tests for when the resolve hook returns `null`? I would prefer if we try to make the `typeof output.then === 'function'` case throw only if we are sure that it was not the user intent (e.g. only throw when `url` is not provided because it's required anyway):
```suggestion
        output === null ||
        (output.url == null && typeof output.then === 'function')
```